src/Symfony/Component/Utf8/Bytes.php

```diff
@@ -0,0 +1,505 @@
+        }
+
+        if (false === $part = strstr($this->string, $needle, $beforeNeedle)) {
+            return;
+        }
+
+        $result = clone $this;
+        $result->string = $part;
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function substringOfIgnoreCase(string $needle, bool $beforeNeedle = false)
+    {
+        if ('' === $needle) {
+            return;
+        }
+
+        $current = setlocale(LC_CTYPE, '0');
+        setlocale(LC_CTYPE, 'C');
+        $part = stristr($this->string, $needle, $beforeNeedle);
+        setlocale(LC_CTYPE, $current);
+
+        if (false === $part) {
+            return;
+        }
+
+        $result = clone $this;
+        $result->string = $part;
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function lastSubstringOf(string $needle, bool $beforeNeedle = false)
+    {
+        if ('' === $needle) {
+            return;
+        }
+
+        if (!$beforeNeedle) {
+
+            if (false === $part = strrchr($this->string, $needle)) {
+                return;
+            }
+
+            $result = clone $this;
+            $result->string = $part;
+
+            return $result;
+        }
+
+        if (false === $offset = strrpos($this->string, $needle)) {
+            return;
+        }
+
+        $result = clone $this;
+        $result->string = substr($this->string, 0, $offset);
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function lastSubstringOfIgnoreCase(string $needle, bool $beforeNeedle = false)
+    {
+        if ('' === $needle) {
+            return;
+        }
+
+        if (false === $offset = strripos($this->string, $needle)) {
+            return;
+        }
+
+        $result = clone $this;
+
+        $current = setlocale(LC_CTYPE, '0');
+        setlocale(LC_CTYPE, 'C');
+        $result->string = $beforeNeedle ? substr($this->string, 0, $offset) : substr($this->string, $offset);
+        setlocale(LC_CTYPE, $current);
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function reverse()
+    {
+        $result = clone $this;
+        $result->string = strrev($this->string);
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function width(bool $ignoreAnsiDecoration = true): int
+    {
+        $s = $this->string;
+
+        if (false !== strpos($s, "\r")) {
+            $s = str_replace("\r\n", "\n", $s);
+            $s = strtr($s, "\r", "\n");
+        }
+
+        $width = 0;
+        foreach (explode("\n", $s) as $s) {
+            if ($ignoreAnsiDecoration) {
+                $s = preg_replace('/\x1B\[[\d;]*m/', '', $s);
+            }
+
+            if ($width < $c = strlen($s)) {
+                $width = $c;
+            }
+        }
+
+        return $width;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function replace($from, $to, int &$count = null)
```

Comment:
To elaborate further, there are many other methods in these classes that take additional flag arguments, and make the API harder to use, e.g. `substringOf('foo', true)` translates to _“what was that `true` doing again?”_ We just had a discussion on internals that did not end nicely regarding the same thing.

A nice way of solving such issues is to return an object with which the result can be further restricted, e.g.:

```php
$str->ignoreCase()->substring('foo')->beforeNeedle();
```

In this case `ignoreCase` would return a decorator that makes all methods (where it makes sense) case insensitive and `substring` would return a result object (just like I proposed for `replace`) that has additional methods, or is actually a dedicated special case result object with a limit API where the caller must invoke `toString` again to unwrap the underlying actual string.

Such APIs are less prone to errors, allow easy discovery of legal actions, and support future extension without BC concerns.

That being said, they introduce a performance penalty in an interpreted language like PHP, for obvious reasons. However, if this stuff becomes a bottleneck, native functions are required, and not faster wrappers. 😉